### DISCUSSION TITLE
Parser

### DISCUSSION THREAD

**Discussion by meta-circular** [biwascheme/biwascheme#293] [General]
https://github.com/biwascheme/biwascheme/blob/038dbcf9c0c0824675a3a07cdd3e53b05fe71cd3/src/system/parser.js#L41

Hello,

I really like the clean naming and implementation  based on R. Kent Dybvig papers chosen by BiwaScheme.
However I have been surprised by the fact there is no info for error locations or that I got strange errors for code like (console-log "Hello")), with the last parenthesis causing an "Unbound symbol )".

Are you using Regex on purpose for performances over a reliable parser ?

**Comment by yhara** [comment]
No, it is just taken from jsScheme and not so much changed for a long time.

**Comment by jcubic** [comment]
If you want something that can be way easier modified and updated for different syntax. You can look at [LIPS Lexer](https://github.com/jcubic/lips/blob/master/src/lips.js#L995) it's a simple state machine for different tokens, the parser was inspired by BiwaScheme. (Unfortunately, I was not able to split the project into modules with too many cycles when importing modules).

In my project, the lexer can be modified by the parser because I have a feature called syntax extensions it's code executed at parse time where you can modify the parser while it's parsing the code (you add new special forms like `#` for vectors). But this is optional and you can easily return an array of tokens from Lexer to process by the parser. This is handled by function [tokens](https://github.com/jcubic/lips/blob/master/src/lips.js#L583) used by tokenized in different places (e.g. auto indentation and formatting).
